[GSoC] Parallelisation of AnalysisBase with multiprocessing and dask (#4162)
MDAnalysis/mdanalysis: package/doc/sphinx/source/documentation_pages/analysis/parallelization.rst

.. -*- coding: utf-8 -*-

=================
Parallel analysis
=================

Starting v.2.8.0, MDAnalysis has introduced process-based parallelization for
virtually all analysis classes. This page will briefly explain what has changed
in the :class:`MDanalysis.analysis.base.AnalysisBase` protocol, how it affects
users and developers, when you should use parallelization (almost always!), and

> when you can abstain from doing so (rarely).

Reviewer: ```suggestion
Starting from v2.8.0, MDAnalysis has introduced process-based parallelization for
virtually all analysis classes. This page will briefly explain what has changed
in the :class:`MDanalysis.analysis.base.AnalysisBase` protocol, how it affects
users and developers, when you should use parallelization (almost always!), and
when you should abstain from doing so (rarely).
```

All this work was done within Google Summer of Code summer project by @marinegor

> and MDAnalysis core developers team.

Reviewer: ```suggestion
All this work was done as part of Google Summer of Code 2023 by @marinegor
and MDAnalysis GSoC mentors.
```

How to use parallelization
==========================

In order to use parallelization in a built-in analysis class ``SomeClass``, simply check which
backends are available, and then just enable them by providing
``backend='multiprocessing'`` and ``n_workers=...`` to ``SomeClass.run(...)`` method:

.. code-block:: python

    u = mda.Universe(...)
    my_run = SomeClass(trajectory)
    assert SomeClass.get_supported_backends() == ('serial', 'multiprocessing', 'dask')

    my_run.run(backend='multiprocessing', n_workers=12)

> For some classes, such as :class:`MDAnalysis.analysis.rms.RMSF`,

Reviewer 2: To be precise, RMSF can be parallelized with split-apply-combine---with a different algorithm (https://www.mdanalysis.org/pmda/api/rmsf.html). I would be more careful with the words here.

Author: Yes, indeed, my bad, I meant a different thing. I assume saying "parallelization of RMSF (in its current implementation) isn't possible" is correct, right?

split-apply-combine parallelization isn't possible, and running them will be
impossible with any but ``serial`` backend.

Also, backends are getting added to the new classes slowly -- in 2.8.0 only
:class:`MDAnalysis.analysis.rms.RMSD` got new backends. But since adding
parallelization to a class is very simple, it won't take much time untill it's

> introduced to all trivial classes.

Reviewer: ```suggestion
Also, backends are getting added to the new classes slowly -- in 2.8.0 only
:class:`MDAnalysis.analysis.rms.RMSD` got new backends. But since adding
parallelization to a class is very simple, it won't take much time until it's
introduced to all trivial classes.
```

How does parallelization work
=============================

The main idea behind its current version is that a trajectory analysis is almost
always trivially parallelizable, meaning you can analyze all frames
independently, and then merge them in a single object. This approach is also
known as "split-apply-combine", and isn't new to MDAnalysis users, since it was
first introduced in [pmda](https://github.com/mdanalysis/pmda). Version 2.8.0 of

> MDAnalysis brings this approach to the main library.

Reviewer: ```suggestion
The main idea behind its current implementation is that a trajectory analysis is
often trivially parallelizable, meaning you can analyze all frames
independently, and then merge them in a single object. This approach is also
known as "split-apply-combine", and isn't new to MDAnalysis users, since it was
first introduced in [pmda](https://github.com/mdanalysis/pmda). Version 2.8.0 of
MDAnalysis brings this approach to the main library.
```

split-apply-combine
-------------------

The following scheme explains the current ``AnalysisBase.run()`` protocol
(user-implemented methods are highlighted in orange):

.. figure:: /images/AnalysisBase_parallel.png

> In short, after checking input parameters and configuring backend,

Reviewer: ```suggestion
In short, after checking input parameters and configuring the backend,
```

> ``AnalysisBase`` splits all the frames into *computation_groups* (equally sized

Reviewer: ```suggestion
``AnalysisBase`` splits all the frames into *computation groups* (equally sized
```
I think it's better to use plain English, unless you use code formatting.

> sequencial groups of frames to be processed by each worker). All groups then get

Reviewer: ```suggestion
sequential groups of frames to be processed by each worker). All groups then get
```

frames of a computation group. Within this method, a user-implemented
``_single_frame`` method gets **applied** to each frame in a computation group.
After that, the main instance gets an object that will **combine** all the
objects from remote workers, and all instances get *merged* with an instance of
:class:`MDAnalysis.analysis.results.ResultsGroup`. Then, a normal
user-implemented ``_compute`` method is called.

Since one of the goals of the project was to **not** break any existing code,
the protocol mimics the single-process parallelization where possible. Meaning,
user-implemented methods such as ``_prepare``, ``_single_frame`` and
``_conclude`` won't need to know they are operating on an instance within the
main python process, or on a remote instance, since the executed code is the
same in both cases.

New methods in ``AnalysisBase``
-------------------------------

From a user point of new, there are no new (non-trivial) methods. Meaning, if
you want to write your own analysis class, you still have to implement only
``_prepare``, ``_single_frame`` and ``_conclude``. However, from a developer point of

> view, there are few:

Reviewer: ```suggestion
From a user point of new, there are no new (non-trivial) methods. If
you want to write your own analysis class, you still have to implement only
``_prepare``, ``_single_frame`` and ``_conclude``. However, from a developer point of
view, there are a few new methods:
```

#. :meth:`MDAnalysis.analysis.base.AnalysisBase._define_run_frames`
#. :meth:`MDAnalysis.analysis.base.AnalysisBase._prepare_sliced_trajectory`
#. :meth:`MDAnalysis.analysis.base.AnalysisBase._configure_backend`
#. :meth:`MDAnalysis.analysis.base.AnalysisBase._setup_computation_groups`
#. :meth:`MDAnalysis.analysis.base.AnalysisBase._compute`

> #. :meth:`MDAnalysis.analysis.base.AnalysisBase._get_aggregator`

Reviewer 3: should `_get_checks` and `_get_warnings` be described here?

Author: Isn't this dedicated to the `AnalysisBase` class?
However, I've added a short section about `_get_checks` and `_get_warnings` below.

Reviewer 3: ah yes, thanks!

> First two methods share the functionality of ``_setup_frames``.

Reviewer: ```suggestion
The first two methods share the functionality of ``_setup_frames``.
```

``_define_run_frames`` is run once during analysis, as it checks that input
parameters ``start, stop, step`` or ``frames`` are consistent with the given
trajectory and prepares the ``slicer`` object that defines the iteration pattern

> through trajectory. ``_prepare_sliced_trajectory`` assigns

Reviewer: ```suggestion
through the trajectory. ``_prepare_sliced_trajectory`` assigns
```

``self._sliced_trajectory`` attribute, and also number of frames in it, and
``self.frames`` and ``self.times`` arrays. In case the computation will be later
split between other processes, this method will be called again on each of the
computation groups.

Method ``_configure_backend`` performs basic health checks for a given analysis

> class -- namely, it compares given backend (if it's a ``str`` instance, such as

Reviewer: ```suggestion
class -- namely, it compares a given backend (if it's a ``str`` instance, such as
```
Or "the"?

``'multiprocessing'``) with the list of builtin backends (and also backends
implemented for a given analysis subclass), and configures a

> :class:`MDAnalysis.analysis.backends.BackendBase` instance accordingly. If user

Reviewer: ```suggestion
:class:`MDAnalysis.analysis.backends.BackendBase` instance accordingly. If the user
```

decides to provide a custom backend (any subclass of
:class:`MDAnalysis.analysis.backends.BackendBase`, or anything with ``apply``

> method), ensures that number of workers wasn't specified twice (during backend

Reviewer: ```suggestion
method), it ensures that number of workers wasn't specified twice (during backend
```

initialization and in ``run()`` arguments).

> After backend is configured, ``_setup_computation_groups`` splits the frames

Reviewer: ```suggestion
After a backend is configured, ``_setup_computation_groups`` splits the frames
```
Or "the"?

> prepared earlier in ``self._prepare_sliced_trajectory`` into number of groups,

Reviewer: ```suggestion
prepared earlier in ``self._prepare_sliced_trajectory`` into a number of groups,
```

by default equal to the number of workers. 

In the ``_compute`` method, frames get initialized again with

> ``_prepare_sliced_trajectory``, and also attributes necessary for a specific

Reviewer: ```suggestion
``_prepare_sliced_trajectory``, and attributes necessary for a specific
```

the :class:`MDAnalysis.analysis.results.Results` class. Since the ``Results``
class is basically a dictionary that also keeps track of assigned attributes, it
is possible to iterate over all these attributes later. ``ResultsGroup`` does
exactly that: given a list of the ``Results`` objects with the same attributes,
it applies a specific aggregation function to every attribute, and stores it as
a same attribute of the returned object:

.. code-block:: python

    from MDAnalysis.analysis.results import ResultsGroup, Results
    group = ResultsGroup(lookup={'mass': ResultsGroup.float_mean})
    obj1 = Results(mass=1)
    obj2 = Results(mass=3)
    assert group.merge([obj1, obj2]) == Results(mass=2.0)

``BackendBase``
---------------

:class:`MDAnalysis.analysis.backends.BackendBase` holds all backend attributes,
and also implements an :meth:`MDAnalysis.analysis.backends.BackendBase.apply`
method, applying a given function to a list of its parameters, but in a parallel
fashion. Although in ``AnalysisBase`` it is used to apply a ``_compute``
function, in principle it can be used to any arbitrary function and arguments,
given they're serializable.

> When to use parallelization *aka* known limitations
> ===================================================

Reviewer: ```suggestion
When to use parallelization? (Known limitations)
================================================
```

For now, the syntax for running parallel analysis is explicit, meaning by
default the ``serial`` version will be run, and the parallelization won't be
enabled by default. Although we expect the parallelization to be useful in most
cases, there are some known caveats from the inital benchmarks.

Fast ``_single_frame`` compared to reading from disk

> --------------------------------------------------

Reviewer 2: ```suggestion
----------------------------------------------------
```

> In all cases, parallelization will be useful only when frames are being

Reviewer: ```suggestion
In all cases, parallelization will not be useful only when frames are being
```
?

Author: ah indeed

processed faster than being read from the disk, otherwise reading is the
bottleneck here. Hence, you'll benefit from parallelization only if you have
relatively much compute per frame, or a fast drive, as illustrated below:

.. figure:: /images/parallelization_time.png

In other words, if you have *fast* analysis (say,
:class:`MDAnalysis.analysis.rms.RMSD`) **and** a slow HDD drive, you are likely

> to not get any benefits from parallelization. Otherwise, you'll be fine.

Reviewer: ```suggestion
to not get any benefits from parallelization. Otherwise, you should be fine.
```

> Seriralization issues
> ---------------------

Reviewer: ```suggestion
Serialization issues
--------------------
```

analysis class that e.g. stores a non-serializable object in one of its
attributes, you might get a serialization error (``PicklingError`` if you're
using a ``multiprocessing`` backend). If you want to get around that, we suggest
trying ``backend='dask'`` (it uses ``dask`` serialization engine instead of
``pickle``).

Out of memory issues
--------------------

If you have large memory footprint of each worker, you can run into
out-of-memory errors (i.e. your server freezes when executing a run). In this
case we suggest decreasing the number of workers from all available CPUs (that
you can get with ``multiprocessing.cpu_count()``) to a smaller number.

Progress bar is missing
-----------------------

It is yet not possible to get a progress bar running with any parallel backend.
If you want an ETA of your analysis, we suggest running it in ``serial`` mode
for the first 10-100 frames with ``verbose=True``, and then running it with
multiple workers. Processing time scales almost linearly, so you can get your
ETA by dividing ``serial`` ETA by the number of workers.

Adding parallelization to your own analysis class
=================================================

> If you want to add parallelization to your own analysis class, first make sure
> your algorithm allows you to do that, i.e. you process each frame independently.

Reviewer: ```suggestion
If you want to add parallelization to your own analysis class, first make sure
your algorithm allows you to do that, i.e. you can process each frame independently.
```

Then it's rather simple -- let's look at the actual code that added
parallelization to the :class:`MDAnalysis.analysis.rms.RMSD`:

.. code-block:: python

    from MDAnalysis.analysis.base import AnalysisBase
    from MDAnalysis.analysis.results import ResultsGroup

    class RMSD(BackendBase):
        @classmethod

>         def get_supported_backends(cls):
>         return ('serial', 'multiprocessing', 'dask',)

Reviewer: ```suggestion
        def get_supported_backends(cls):
            return ('serial', 'multiprocessing', 'dask',)
```

        @classmethod
        def is_parallelizable(self):
            return True
        
        def _get_aggregator(self):
            return ResultsGroup(lookup={'rmsd': ResultsGroup.ndarray_vstack})

> That's it! First two methods are boilerplate -- ``get_supported_backends``

Reviewer: ```suggestion
That's it! The first two methods are boilerplate -- ``get_supported_backends``
```

``backend='multiprocessing'`` and ``n_workers=...`` to ``SomeClass.run(...)`` method:

.. code-block:: python

    u = mda.Universe(...)
    my_run = SomeClass(trajectory)
    assert SomeClass.get_supported_backends() == ('serial', 'multiprocessing', 'dask')

    my_run.run(backend='multiprocessing', n_workers=12)

For some classes, such as :class:`MDAnalysis.analysis.rms.RMSF` (in its current implementation),
split-apply-combine parallelization isn't possible, and running them will be
impossible with any but ``serial`` backend.

Also, backends are getting added to the new classes slowly -- in 2.8.0 only
:class:`MDAnalysis.analysis.rms.RMSD` got new backends. But since adding
parallelization to a class is very simple, it won't take much time until it's
introduced to all trivial classes.

How does parallelization work
=============================

The main idea behind its current implementation is that a trajectory analysis is
often trivially parallelizable, meaning you can analyze all frames
independently, and then merge them in a single object. This approach is also
known as "split-apply-combine", and isn't new to MDAnalysis users, since it was

> first introduced in [pmda](https://github.com/mdanalysis/pmda). Version 2.8.0 of

Reviewer 4: fix link as [inline reST link](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#external-links)
```suggestion
first introduced in `PMDA <https://github.com/mdanalysis/pmda>`_. Version 2.8.0 of
```

.. -*- coding: utf-8 -*-

.. _parallel-analysis:
   
=================
Parallel analysis
=================

.. versionadded:: 2.8.0
   Parallelization of analysis classes was added during Google Summer of Code

>    2023 by [@marinegor](https://github.com/marinegor) and MDAnalysis GSoC mentors.

Reviewer 4: reST not markdown
```suggestion
   2023 by `@marinegor <https://github.com/marinegor>`_ and MDAnalysis GSoC mentors.
```

.. -*- coding: utf-8 -*-

.. _parallel-analysis:
   
=================
Parallel analysis
=================

.. versionadded:: 2.8.0
   Parallelization of analysis classes was added during Google Summer of Code

>    2023 by @marinegor and MDAnalysis GSoC mentors.

Reviewer: ```suggestion
   2023 by @marinegor and MDAnalysis GSoC mentors.
```
Maybe add an actual link to your GitHub profile? (I don't think this is linked automatically, but might be wrong.)

Author: Added, thanks. I think you're right, it's not linked automatically (at least online or in vscode)

Starting with v2.8.0, MDAnalysis, introduces process-based parallelization for

> analysis classes. This section explains the implementation and background for

Reviewer 5: ```suggestion
This section explains the implementation and background for
```
That's explicit in the `versionadded` above.

parallelization with the :class:`MDAnalysis.analysis.base.AnalysisBase`, what 
users and developers need to know, when you should use parallelization (almost 
always!), and when you should abstain from doing so (rarely).

How to use parallelization
==========================

> In order to use parallelization in a built-in analysis class ``SomeClass``,

Reviewer: ```suggestion
In order to use parallelization in a built-in analysis class ``SomeAnalysisClass``,
```
Nitpick, feel free to ignore. It would need to be changed elsewhere.

simply check which backends are available (see :ref:`backends` for backends
that are generally available), and then just enable them by providing
``backend='multiprocessing'`` and ``n_workers=...`` to ``SomeClass.run(...)``
method:

.. code-block:: python

    u = mda.Universe(...)
    my_run = SomeClass(trajectory)
    assert SomeClass.get_supported_backends() == ('serial', 'multiprocessing', 'dask')

    my_run.run(backend='multiprocessing', n_workers=12)

For some classes, such as :class:`MDAnalysis.analysis.rms.RMSF` (in its current implementation),
split-apply-combine parallelization isn't possible, and running them will be
impossible with any but the ``serial`` backend.

.. Note::

   Parallelization is getting added to existing analysis classes.  Initially,
   only :class:`MDAnalysis.analysis.rms.RMSD` supports parallel analysis, but
   the goal is that the 3.0 release will support parallel analysis where

>    possible.

Reviewer 5: ```suggestion
   only :class:`MDAnalysis.analysis.rms.RMSD` supports parallel analysis, but
   we aim to increase support in future releases.
```
I'd prefer we don't promise extra things for 3.0, especially when it's not linked to an API break. cc @orbeckst

Author: Didn't resolve or commit that -- leave it up to @orbeckst

first introduced in `PMDA`_ :footcite:p:`Fan2019`.  
Version 2.8.0 of MDAnalysis brings this approach to the main library.

.. _`PMDA`: https://github.com/mdanalysis/pmda

split-apply-combine
-------------------

The following scheme explains the current :meth:`AnalysisBase.run
<MDAnalysis.analysis.base.AnalysisBase.run>` protocol (user-implemented methods
are highlighted in orange):

.. figure:: /images/AnalysisBase_parallel.png

In short, after checking input parameters and configuring the backend,
:class:`AnalysisBase <<MDAnalysis.analysis.base.AnalysisBase>` splits all the
frames into *computation groups* (equally sized sequential groups of frames to
be processed by each worker). All groups then get **split** between workers of
a backend configured early, the main instance gets serialized and distributed
between workers, and then
:meth:`~MDAnalysis.analysis.base.AnalysisBase._compute()` method gets called
for all frames of a computation group. Within this method, a user-implemented
:meth:`~MDAnalysis.analysis.base.AnalysisBase._single_frame` method gets
**applied** to each frame in a computation group.  After that, the main

> instance gets an object that will **combine** all the objects from remote

Reviewer: ```suggestion
instance gets an object that will **combine** all the objects from other
```
I think "remote" only applied to `dask-distributed` that was finally dropped for this PR?

Progress bar is missing
-----------------------

It is yet not possible to get a progress bar running with any parallel backend.
If you want an ETA of your analysis, we suggest running it in ``serial`` mode
for the first 10-100 frames with ``verbose=True``, and then running it with
multiple workers. Processing time scales almost linearly, so you can get your
ETA by dividing ``serial`` ETA by the number of workers.

.. _adding-parallelization:

Adding parallelization to your own analysis class
=================================================

If you want to add parallelization to your own analysis class, first make sure
your algorithm allows you to do that, i.e. you can process each frame independently.
Then it's rather simple -- let's look at the actual code that added
parallelization to the :class:`MDAnalysis.analysis.rms.RMSD`:

.. code-block:: python

    from MDAnalysis.analysis.base import AnalysisBase
    from MDAnalysis.analysis.results import ResultsGroup

    class RMSD(BackendBase):
        @classmethod
        def get_supported_backends(cls):
            return ('serial', 'multiprocessing', 'dask',)

        @classmethod
        def is_parallelizable(self):

>             return True

Reviewer 4: Propose to make this an attribute.

Author: See discussion above
